### PR TITLE
Support i less than j-1 in 2 qubit gates

### PR DESCRIPTION
**Pull request by odanielson**
I think we need to support CNOT operations where ctrl and target qubits are not adjacent qubits. 
For example CNOT q0 q5. 
This PR solves it by 

1. Swap up q0 to q4
2. Apply CNOT on q4 q5
3. Swap down q4 to q0 again

The result from the examples/cnot_test.qc that applies CNOT on x0 (qubit 0) and w0 (qubit 3) is

    Program: Seq([X ['x0'], H ['y0'], H ['z0'], CNOT ['x0', 'w0']])
    Initial state: 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0
    Final state: 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.25, 0.0, 0.25, 0.0, 0.25, 0.0, 0.25 
    Qubit measure: 1, 0, 1, 1

### CODE REVIEW

File: examples/cnot_test.qc [hunk 1]
```diff
@@ -0,0 +1,10 @@
+
+register x0[0]
+register y0[1]
+register z0[2]
+register w0[3]
+
+X x0
+H y0
+H z0
+CNOT x0 w0
```
Comment (DanielJanzon) [inline examples/cnot_test.qc]: I don't understand the purpose of this example...

File: gatearray.py [hunk 2]
```diff
@@ -9,16 +9,17 @@
...
 
 
-START = namedtuple('START', 'n')
-X = namedtuple('X', 'i')
 CNOT = namedtuple('CNOT', 'ctrl target')
 H = namedtuple('H', 'i')
 I = namedtuple('I', 'i')
-
+START = namedtuple('START', 'n')
+SWAP = namedtuple('SWAP', 'a b')
```
Comment (DanielJanzon) [inline gatearray.py]: We might want to start building a macro library instead of expanding the set of gates.

File: gates.py [hunk 3]
```diff
@@ -67,3 +73,9 @@ def __str__(self):
                        [0.0, 1.0, 0.0, 0.0],
                        [0.0, 0.0, 0.0, 1.0],
                        [0.0, 0.0, 1.0, 0.0]], dtype=np.complex256))
+
+SWAP = Gate('SWAP',
+            np.matrix([[1.0, 0.0, 0.0, 0.0],
+                       [0.0, 0.0, 1.0, 0.0],
+                       [0.0, 1.0, 0.0, 0.0],
+                       [0.0, 0.0, 0.0, 1.0]], dtype=np.complex256))
```
Comment (DanielJanzon) [inline gates.py]: Seems correct to me but it's worthy of short comment why this matrix is true. Something like 
```
# (a b) tensor (c d) = (ac bc ad bd). By swapping the middle two elements, producing 
# (ac ad bc bd) we can se by inspection that this is the result of (c d) tensor (a b).
# Hence we arrive at the matrix below that swap the middle elements of the 4-vector.
```
Or is there a more obvious way to realize the form of the swap matrix?

File: simulator.py [hunk 4]
```diff
@@ -25,11 +26,22 @@ def expand_single_gate(gate, i, num_qbits):
...
+    right_bits = num_qbits - 2 - k
+
+    operator = (Identity ** left_bits) * gate * (Identity ** right_bits)
+
+    # Swap left bit until it is left of right bit before real gate operator
+    # and swap it left again after real gate operator
+    while k > i:
+        swap_operator = expand_double_gate(SWAP, k-1, k, num_qbits)
+        operator = swap_operator | operator | swap_operator
+        k -= 1
```
Comment (DanielJanzon) [inline simulator.py]: The overall idea looks correct but the risk of getting a one-off error is big. We should add a unit test of some kind.

File: gatearray.py (hunk 2 above)
Comment (odanielson) [reply gatearray.py]: Might be, but I'll merge this for now so we can write qcode with CNOT operations on non adjacent qubits.

File: simulator.py (hunk 4 above)
Comment (odanielson) [reply simulator.py]: added a unittest :)

**Pull request merged**